pkg/disk/partition_table.go

```diff
@@ -160,6 +161,146 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// partitions specify a type.
+	// The default type is also used for the automatically created /boot
+	// filesystem if it is a supported type for that fileystem. If it is not,
+	// xfs is used as a fallback.
+	DefaultFSType FSType
+
+	// RequiredMinSizes defines a map of minimum sizes for specific
+	// directories. These indirectly control the minimum sizes of partitions. A
+	// directory with a required size will set the minimum size of the
+	// partition with the mountpoint that contains the directory. Additional
+	// directory requirements are additive, meaning the minimum size for a
+	// mountpoint's partition is the sum of all the required directory sizes it
+	// will contain.
+	RequiredMinSizes map[string]uint64
+}
+
+// Returns the default filesystem type if the fstype is empty. If both are
+// empty/none, returns an error.
+func (options *CustomPartitionTableOptions) getfstype(fstype string) (string, error) {
+	if fstype != "" {
+		return fstype, nil
+	}
+
+	if options.DefaultFSType == FS_NONE {
+		return "", fmt.Errorf("no filesystem type defined and no default set")
+	}
+
+	return options.DefaultFSType.String(), nil
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	if options == nil {
+		// init options with empty struct to use defaults
+		options = &CustomPartitionTableOptions{
+			PartitionTableType: PT_GPT,
+		}
+	}
+
+	if customizations == nil {
+		customizations = &blueprint.PartitioningCustomization{}
+	}
+
+	// validate the partitioning customizations before using them
+	if err := customizations.Validate(); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+
+	pt := &PartitionTable{}
+
+	switch options.PartitionTableType {
+	case PT_GPT, PT_DOS:
+		pt.Type = options.PartitionTableType.String()
+	case PT_NONE:
+		// default to "gpt"
+		pt.Type = PT_GPT.String()
+	default:
+		return nil, fmt.Errorf("error generating partition table: invalid partition table type enum value: %d", options.PartitionTableType)
+	}
+
+	// The boot type will be the default only if it's a supported filesystem
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	// FS_NONE also falls back to xfs.
+	var bootType FSType
+	switch options.DefaultFSType {
+	case FS_EXT4, FS_XFS:
+		bootType = options.DefaultFSType
+	default:
+		bootType = FS_XFS
+	}
+
+	if err := addPartitionsForBootMode(pt, options.BootMode); err != nil {
+		return nil, fmt.Errorf("error generating partition table: %w", err)
+	}
+	if customizations.LVM != nil || customizations.Btrfs != nil {
+		// we need a /boot partition to boot LVM or Btrfs, create boot
+		// partition if it does not already exist
+		ensureBoot(pt, bootType)
```

Comment:
(super nitpick) the name `ensureBoot` breaks the pattern of `addPartitionsFor*` that the we here maybe something `ensureBootPartition()` or `addBootPartition()` to be slightly more similar to the other names?

Reply:
`ensureBootPartition()` works.  I wouldn't want to call it `addBootPartition()` though.  I think using `ensure` in these functions (for root too) is a good way to imply that it might not create anything (if it already exists), whereas the `add...` functions always create things (assuming the args aren't nil).

Reply:
Renamed.